### PR TITLE
Spec the constructor string parser.

### PR DESCRIPTION
**Pull request by wanderview**
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/urlpattern/pull/69.html" title="Last updated on Aug 12, 2021, 3:13 PM UTC (493de8e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/urlpattern/69/26710dc...493de8e.html" title="Last updated on Aug 12, 2021, 3:13 PM UTC (493de8e)">Diff</a>

### CODE REVIEW

**Comment by wanderview** [comment]
@domenic PTAL, although it can wait until after the other PRs are merged.  I think this is the last substantive PR.  After this I think it will mostly be polish, non-normative prose, etc.

File: spec.bs
Comment (domenic) [inline spec.bs:468]:
```suggestion
  <p class=allow-2119>Finally, the URLPattern constructor string parser does not handle some parts of the [=basic URL parser=] state machine.  For example, it does not treat backslashes specially as they would all be treated as pattern characters and would require excessive escaping.  In addition, this parser may not handle some more esoteric parts of the URL parsing algorithm like file URLs with a hostname.  The goal with this parser was to handle the most common URLs while allowing any niche case to be handled instead via the {{URLPatternInit}} constructor.
```

File: spec.bs
Comment (domenic) [inline spec.bs:482]:
```suggestion
      <p class="note allow-2119">On every iteration of the parse loop the |parser|'s [=constructor string parser/token index=] will be incremented by its [=constructor string parser/token increment=] value.  Typically this means incrementing by 1, but at certain times it is set to zero.  The [=constructor string parser/token increment=] is then always reset back to 1 at the top of the loop.
```

File: spec.bs
Comment (domenic) [inline spec.bs:493]:
```suggestion
          1. Set |parser|'s [=constructor string parser/result=]["{{URLPatternInit/hash}}"] to the empty string.
```

same for several other cases.

File: spec.bs
Comment (domenic) [inline spec.bs:557]: Nit: I personally like splitting these up, i.e.

> Else if any of the following are true:
> 
> * lowercase sentence fragment;
> * lowercase sentence fragment; or
> * lowercase sentence fragment,
> 
> then run...

File: spec.bs
Comment (domenic) [inline spec.bs:611]: This only has the previous algorithm as a caller, so probably is best inlined into it?

File: spec.bs
Comment (domenic) [inline spec.bs:650]: See my above nit about style, but even if you don't go that route, this should be a ul, not ol.

File: spec.bs
Comment (domenic) [inline spec.bs:816]: Good catch!

**Review by domenic — APPROVED** [review]
Although I can't claim to have run the algorithms in my head and verified they're correct, overall it looks pretty solid! Only a few editorial suggestions and nits.

File: spec.bs
Comment (wanderview) [reply spec.bs:611]: Huh, its not called separately in the c++ either.  It definitely was before.  I must have gotten rid of that callsite in a refactor at some point.  I'll inline here.

**Pull request merged**
